File: week1/1-Vector/vector.py
class Vector:

    def __init__(self):
        self.__size = 0
        self.__capacity = 10
        self.elements = self.__capacity * [None]


    def insert(self, index, value):
        """
        Adds value at a specific index in the Vector.
        Complexity: O(n)
        """
        if self.__size == self.__capacity:
            self.expand()

        if index > self.__size:
            index = self.__size

        self.elements[index:index] = [value]
        self.__size += 1


    def add(self, value):
        """
        Adds value to the end of the Vector.
        Complexity: O(1)
        """
        if self.__size == self.__capacity:
            self.expand()

        self.elements[self.__size] = value
        self.__size += 1


    def expand(self):
        """
        Increases the __capacity of elements
        """
        self.elements.extend(self.__capacity * [None])
        self.__capacity *= 2


    def get(self, index):
        """
        Returns value at a specific index in the Vector
        Complexity: O(1)
        """
        return self.elements[index]


    def remove(self, index):
        """
        Removes element at the specific index
        Complexity: O(n)
        """
        if self.__size:
            del self.elements[index]
            self.__size -= 1


    def pop(self):
        """
        Removes element at the last index
        Complexity: O(1)
        """
        if self.__size:
            del self.elements[self.__size - 1]
            self.__size -= 1


    def size(self):
        """
        Returns the number of elements in the Vector.
        Complexity: O(1)
        """
        return self.__size


    def capacity(self):
        """
        Returns the total __capacity of the Vector.
        Complexity: O(1)
        """
        return self.__capacity


    def print(self):
        """
        Prints elements
        """
        print(self.elements[:self.__size])

File: week1/1-Vector/Vector.java

/**
 *
 * @author qvka
 */
public class Vector<T> {
    private T[] array;
    private int size;
    public Vector(){
        this.size = 0;
        this.array = (T[]) new Object[8];
    }
    public Vector(int capacity){
        this.size = 0;
        this.array = (T[]) new Object[capacity];
    }
    // Adds value at a specific index in the Vector.
    // Complexity: O(n)
    public void insert(int index, T value) {
        if(index > size || index < 0){
            throw new IndexOutOfBoundsException();
        }
        if(size == array.length){
            expand();
        }
        int i = size;
       
        while( i > index ){
            array[i] = array[i-1];
            i--;
        }
        array[index] = value;
        size++;
    }

    // Adds value to the end of the Vector.
    // Complexity: O(1)
    public void add(T value) {
        if(size == array.length){
            expand();
        }
        array[size] = value;
        size++;
    }
    private void expand(){
        T[] temp  = (T[]) new Object[array.length*2];
        System.arraycopy(array, 0, temp, 0, size);
        array = temp;
    }
    // Returns value at a specific index in the Vector
    // Complexity: O(1)
    public T get(int index) {
        if( index >= size || index < 0){
            throw new IndexOutOfBoundsException();
        }
        return (T)array[index];
    }

    // Removes element at the specific index
    // Complexity: O(n)
    public void remove(int index) {
        if(size == 0){
            return;
        }
        if( index < 0 || index >= size){
            throw new IndexOutOfBoundsException();
        }
        int i = index+1;
        while( i < size){
            array[i-1] = array[i];
            i++;
        }
        size--;
    }

    // Removes element at the last index
    // Complexity: O(1)
    public T pop() {
        if(size == 0){
            return null;
        }
        size--;
        return (T)array[size+1];
    }

    // Returns the number of elements in the Vector.
    // Complexity: O(1)
    public int size() {
        return size;
    }

    // Returns the total capacity of the Vector.
    // Complexity: O(1)
    public int capacity() {
        return array.length;
    }

    public static void main(String[] args) {
        Vector<Integer> a = new Vector(5);
        System.out.println(a.capacity() + " " +a.size());
        a.add(2);
        System.out.println(a.get(0));
        a.insert(0, 1);
        System.out.println(a.get(0));
        a.remove(0);
        a.add(6);
        a.add(6);
        a.add(6);
        a.add(6);
        a.add(6);
        a.add(6);
        System.out.println(a.capacity());
       
    } 
}
